packages/internal-test-utils/internalAct.js

```diff
@@ -194,6 +195,39 @@ export async function act<T>(scope: () => Thenable<T>): Thenable<T> {
   }
 }
 
+async function waitForTasksAndTimers(error: mixed) {
+  do {
+    // Wait until end of current task/microtask.
+    await waitForMicrotasks();
+
+    // $FlowFixMe[cannot-resolve-name]: Flow doesn't know about global Jest object
+    if (jest.isEnvironmentTornDown()) {
+      error.message =
+        'The Jest environment was torn down before `act` completed. This ' +
+        'probably means you forgot to `await` an `act` call.';
+      throw error;
+    }
+
+    // $FlowFixMe[cannot-resolve-name]: Flow doesn't know about global Jest object
+    const j = jest;
+    if (j.getTimerCount() > 0) {
+      // There's a pending timer. Flush it now. We only do this in order to
+      // force Suspense fallbacks to display; the fact that it's a timer
+      // is an implementation detail. If there are other timers scheduled,
+      // those will also fire now, too, which is not ideal. (The public
+      // version of `act` doesn't do this.) For this reason, we should try
+      // to avoid using timers in our internal tests.
+      j.runAllTicks();
```

Comment:
`getTimerCount()` includes `process.nextTick` mocks but `runOnlyPendingTimers()` alone doesn't run them. We don't use it ourselves but JSDOM uses it internally. This was driving me insane debugging this infinite loop.

Reply:
should we `waitForMicrotasks` here before running timers. that probably more closely approximates a real loop

Reply:
could continue the loop if runAllTicks changes the timer count. It's unfortunate you can't know if ticks ran. unless there isa. getTickCount or something

Reply:
Technically yes, but we also don't wait between multiple timers or multiple ticks.

What we really want is the `runAllTimersAsync()` but for whatever reason that's not available on the legacy mocks. I guess we'd have to upgrade to newer mock timers.